File: src/runtime/boxes/BbInterr.java
/**
 * BbInterr.java — INTERR: ?X — succeed zero-width if X succeeds; ω if X fails
 * Port of bb_interr.c / bb_interr.s
 *
 * o$int semantics: run child; if child γ → discard match, return zero-width
 * at the *original* cursor; if child ω → propagate ω.
 * β: unconditional ω — interrogation succeeds at most once per position.
 *
 *   INT_α:  start=Δ; cr=child(α); if empty → INT_ω;
 *           Δ=start; goto INT_γ;
 *   INT_β:            goto INT_ω;
 *   INT_γ:  return spec(Σ+Δ,0);
 *   INT_ω:  return spec_empty;
 */
class BbInterr extends BbBox {
    private final BbBox child;
    private int         start;

    public BbInterr(MatchState ms, BbBox child) { super(ms); this.child=child; }

    @Override public Spec alpha() {
        start    = ms.delta;
        Spec cr  = child.alpha();
        if (cr == null) return null;                                           // INT_ω
        ms.delta = start;                                                      // discard advance
        return new Spec(ms.delta, 0);                                          // INT_γ zero-width
    }

    @Override public Spec beta() { return null; }                              // INT_ω
}

File: src/runtime/boxes/bb_interr.c
/* _XINT     INTERR      ?X — null result if X succeeds; ω if X fails (o$int) */
#include "bb_box.h"
#include <stdlib.h>
#include <string.h>

/* o$int: replace operand with null, continue.
 * In BB terms: run child; if child γ → discard match, return zero-width at
 * the *original* cursor (null string); if child ω → propagate ω.
 * β: unconditional ω — interrogation succeeds at most once. */
typedef struct { bb_box_fn fn; void *state; int start; } interr_t;

spec_t bb_interr(void *zeta, int entry)
{
    interr_t *ζ = zeta;
    spec_t cr;
    if (entry==α)                                                               goto INT_α;
    if (entry==β)                                                               goto INT_β;
    INT_α:          ζ->start=Δ;                                                 \
                    cr=ζ->fn(ζ->state,α);                                       \
                    if (spec_is_empty(cr))                                      goto INT_ω;
                    Δ=ζ->start;                                                 goto INT_γ;
    INT_β:                                                                      goto INT_ω;
    INT_γ:                                                                      return spec(Σ+Δ,0);
    INT_ω:                                                                      return spec_empty;
}

interr_t *bb_interr_new(bb_box_fn fn, void *state)
{ interr_t *ζ=calloc(1,sizeof(interr_t)); ζ->fn=fn; ζ->state=state; return ζ; }
